### PR TITLE
Transient buffers

### PR DESCRIPTION
**Pull request by jrieken**
This PR adds support for handling unknown buffers. The idea goes like this:

1. a ```Request``` for an unknown files comes in and
2. the buffer manager simply adds this file to all projects (I figured it's better then guessing, but we can discuss) and
3. the buffer manager remembers that file path so that later when
4. a project system adds the same file (because it caught up after a FS change event or ```/addToProject``` was called) the buffer manager can remove the transient file again (avoid duplication)

### CODE REVIEW

File: src/OmniSharp/Roslyn/BufferManager.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using OmniSharp.Models;
+
+namespace OmniSharp.Roslyn
+{
+    public class BufferManager
+    {
+
```
Comment (davidfowl) [inline src/OmniSharp/Roslyn/BufferManager.cs]: Got an extra space :smile:

Comment (nosami) [inline src/OmniSharp/Roslyn/BufferManager.cs]: We need a test for that

File: tests/OmniSharp.Tests/UpdateBufferFilterFacts.cs [hunk 2]
```diff
@@ -42,5 +43,62 @@ public class UpdateBufferFilterFacts
...
+            var sourceText = await workspace.CurrentSolution.GetDocument(docId).GetTextAsync();
+            Assert.Equal("interface I {}", sourceText.ToString());
+            
+            docId = workspace.CurrentSolution.GetDocumentIdsWithFilePath("test.cs").FirstOrDefault();
+            Assert.NotNull(docId);
+            sourceText = await workspace.CurrentSolution.GetDocument(docId).GetTextAsync();
+            Assert.Equal("class C {}", sourceText.ToString());
+
+
+            var project1 = workspace.CurrentSolution.Projects.First();
```
Comment (nosami) [inline tests/OmniSharp.Tests/UpdateBufferFilterFacts.cs]: this needs to go :)

**Comment by nosami** [comment]
I've just been trying this out. Seems to work fine :+1:

I did get extra references during intellisense (So I get Xunit references in the OmniSharp project for example) - but I guess you were expecting that. 

What if we had 100 projects? I imagine intellisense would be very slow for the transient file because of all the references. (similar to the dirty hack I originally put into omnisharp-server for vnext that added all references in a folder - DF's seen this)

**Comment by nosami** [comment]
Something is missing somewhere :) 

I can create a test class in the folder with the other tests. I can get intellisense for `using Xunit` and `[Fact]` etc.

But diagnostics shows me `The type or namespace name 'Xunit' could not be found`
But I get diagnostics saying 'The ty

**Comment by jrieken** [comment]
@nosami - good catch. the problem is that the file is now part of all projects and some don't have a dependency onto ```Xunit``` (e.g Omnisharp and Omnisharp.Stdio). That's why you see that error. Seems like my simplification _add file to all projects_ doesn't work and guessing is required. I will push a second change that selects projects based on file path.

**Comment by jrieken** [comment]
@nosami - I have changed the logic to find the best matching projects based on the file path (contains relation). Also added more tests.

**Comment by nosami** [comment]
```
k test --test OmniSharp.Tests.BufferManagerFacts.UpdateBufferIgnoresFilePathsThatDontMatchAProjectPath
xUnit.net Project K test runner (64-bit Asp.Net,Version=v5.0)
Copyright (C) 2014 Outercurve Foundation, Microsoft Open Technologies, Inc.

Starting: OmniSharp.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
   OmniSharp.Tests.BufferManagerFacts.UpdateBufferIgnoresFilePathsThatDontMatchAProjectPath [FAIL]
      Assert.Equal() Failure\nExpected: 0\nActual:   4
      Stack Trace:
           at OmniSharp.Tests.BufferManagerFacts.UpdateBufferIgnoresFilePathsThatDontMatchAProjectPath () [0x00000] in <filename unknown>:0 
           at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
           at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0 
Finished: OmniSharp.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null

=== TEST EXECUTION SUMMARY ===
   OmniSharp.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null  Total: 1, Failed: 1, Skipped: 0, Time: 1.063s

k test --test OmniSharp.Tests.BufferManagerFacts.UpdateBufferFindsProjectBasedOnPath
xUnit.net Project K test runner (64-bit Asp.Net,Version=v5.0)
Copyright (C) 2014 Outercurve Foundation, Microsoft Open Technologies, Inc.

Starting: OmniSharp.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
   OmniSharp.Tests.BufferManagerFacts.UpdateBufferFindsProjectBasedOnPath [FAIL]
      Assert.Equal() Failure\nExpected: 2\nActual:   4
      Stack Trace:
           at OmniSharp.Tests.BufferManagerFacts.UpdateBufferFindsProjectBasedOnPath () [0x00000] in <filename unknown>:0 
           at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
           at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0 
Finished: OmniSharp.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null

k test --test OmniSharp.Tests.BufferManagerFacts.UpdateBufferFindsProjectBasedOnNearestPath
xUnit.net Project K test runner (64-bit Asp.Net,Version=v5.0)
Copyright (C) 2014 Outercurve Foundation, Microsoft Open Technologies, Inc.

Starting: OmniSharp.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
   OmniSharp.Tests.BufferManagerFacts.UpdateBufferFindsProjectBasedOnNearestPath [FAIL]
      Assert.Equal() Failure\nExpected: 1\nActual:   2
      Stack Trace:
           at OmniSharp.Tests.BufferManagerFacts.UpdateBufferFindsProjectBasedOnNearestPath () [0x00000] in <filename unknown>:0 
           at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
           at System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00000] in <filename unknown>:0 
Finished: OmniSharp.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
```

If I run all tests, with `k test`, it hangs forever :S

**Comment by jrieken** [comment]
funky - is it that mono cannot handle the windows file paths?

**Comment by nosami** [comment]
More like linux can't handle Windows paths :) Mono has a shim for Windows compatibility, but it's disabled by default.

```cs
        [Fact]
        public void UpdateBufferIgnoresFilePathsThatDontMatchAProjectPath()
        {
            var workspace = GetWorkspaceWithProjects();

            workspace.BufferManager.UpdateBuffer(new Request() { FileName = @"/some/path.cs", Buffer = "enum E {}" });
            var documents = workspace.GetDocuments(@"/some/path.cs");
            Assert.Equal(0, documents.Count());
        }
```

^^ this passes. Does that work on Windows too? I know Windows can handle forward slash as a path delimiter, but not so sure about omitting the drive letter.

**Comment by nosami** [comment]
I've just been trying the code out. It works perfectly for me :D 

I was able to create a new file with a test and run it - no problems :)

**Comment by jrieken** [comment]
Great. Pushed now a change that uses ```Path.Combine``` and paths without drive letter (and also checked my Mac order)

**Comment by nosami** [comment]
These tests are working now... but `k test` still hangs... no idea why. Does it work locally for you?

**Comment by jrieken** [comment]
yeah - all happy. can you run them locally?

image

**Comment by nosami** [comment]
try the other tests :)

**Comment by jrieken** [comment]
I did. wrong screen shot - fixed inline

**Comment by nosami** [comment]
I can't run them all together with `k test`... it just hangs.

**Comment by jrieken** [comment]
will digg out an Ubuntu box and try there

**Comment by nosami** [comment]
I've fixed.

**Comment by jrieken** [comment]
mind bending

**Comment by jrieken** [comment]
thanks a lot!

**Comment by nosami** [comment]
I can't believe we can actually handle new files now :dancers: Cheers Jo!

**Comment by jrieken** [comment]
:beers: just in time for Friday

**Pull request merged**